encoder.lua

```diff
@@ -67,74 +67,152 @@ local function toms(timestamp)
     return string.format("%.3f", timestamp)
 end
 
-local function quality_to_crf(quality, max_crf)
-    -- Quality is from 0 to 100. (for avif images) CRF is from 0 to 63 and reversed.
-    return math.floor((100 - quality) / 100 * max_crf)
+local function rescale_quality(quality, min_q, max_q)
+    local scaled = min_q + (max_q - min_q) * quality / 100
+    -- Round to the nearest integer that's better in quality.
+    if min_q > max_q then
+        return math.floor(scaled)
+    end
+    return math.ceil(scaled)
 end
 
 ------------------------------------------------------------
 -- ffmpeg encoder
 
 local ffmpeg = {}
 
-ffmpeg.prefix = { find_exec("ffmpeg"), "-hide_banner", "-nostdin", "-y", "-loglevel", "quiet", "-sn", }
+ffmpeg.exec = find_exec("ffmpeg")
 
-ffmpeg.prepend = function(args)
-    if next(args) ~= nil then
-        for i, value in ipairs(ffmpeg.prefix) do
-            table.insert(args, i, value)
-        end
-    end
-    return args
+ffmpeg.prepend = function(...)
+    return {
+        ffmpeg.exec, "-hide_banner", "-nostdin", "-y", "-loglevel", "quiet", "-sn",
+        ...,
+    }
 end
 
 ffmpeg.make_static_snapshot_args = function(source_path, output_path, timestamp)
-    local args = ffmpeg.prepend {
+    local encoder_args
+    if self.config.snapshot_format == 'avif' then
+        encoder_args = {
+            '-c:v', 'libaom-av1',
+            '-cpu-used', '6', -- cpu-used < 6 can take a lot of time to encode.
+            '-crf', tostring(rescale_quality(self.config.snapshot_quality, self.max_avif_crf, 0)),
+            '-still-picture', '1',
+        }
+    elseif self.config.snapshot_format == 'webp' then
+        encoder_args = {
+            '-c:v', 'libwebp',
+            '-compression_level', '6',
+            '-quality', tostring(self.config.snapshot_quality),
+        }
+    else
+        encoder_args = {
+            '-c:v', 'mjpeg',
+            '-q:v', tostring(rescale_quality(self.config.snapshot_quality, 31, 2)),
+        }
+    end
+
+    local args = ffmpeg.prepend(
         '-an',
         '-ss', toms(timestamp),
         '-i', source_path,
         '-map_metadata', '-1',
-        '-vcodec', self.config.snapshot_codec,
-        '-lossless', '0',
-        '-compression_level', '6',
-        '-qscale:v', tostring(self.config.snapshot_quality),
-        '-vf', string.format('scale=%d:%d', self.config.snapshot_width, self.config.snapshot_height),
-        '-vframes', '1',
-        output_path
-    }
-    if self.config.snapshot_format == 'avif' then
-        -- Avif quality can be controlled with crf.
-        table.insert(args, #args, '-crf')
-        table.insert(args, #args, tostring(quality_to_crf(self.config.snapshot_quality, self.max_avif_crf)))
-    end
+        '-vf', string.format("scale='min(%d,iw)':'min(%d,ih)':flags=sinc+accurate_rnd",
+                             self.config.snapshot_width, self.config.snapshot_height),
+        '-frames:v', '1',
+        table.unpack(encoder_args)
```

Comment:
Using table.unpack() crashes mpvacious. 

> attempt to call field 'unpack' (a nil value)

Use h.unpack() instead.

Reply:
Better be commented somewhere.